lom/src/engine/context.go

```diff
@@ -0,0 +1,386 @@
+    tout := A_DAY_IN_SECS
+
+    for {
+        select {
+        case clReq := <- p.pendingReadRequests:
+            w := &wait_t { req: clReq }
+            if clReq.Req.TimeoutSecs == 0 {
+                listNoTimeout = append(listNoTimeout, w)
+            } else {
+                w.due = time.Now().Unix() + int64(clReq.Req.TimeoutSecs)
+                listWTimeout = append(listWTimeout, w)
+                sort.Slice(listWTimeout, func(i, j int) bool {
+                    return listWTimeout[i].due < listWTimeout[j].due
+                })
+            }
+
+        case serReq := <- p.pendingWriteRequests:
+            serverRequests = append(serverRequests, serReq)
+
+        case <- time.After(time.Duration(tout) * time.Second):
+            /* bail out */
+
+        case <- p.abortCh:
+            LogInfo("Aborting Send requests")
+            return
+        }
+        
+        /* Here you come on client/server request or timeout */
+        for len(serverRequests) > 0 {
+            var r *LoMRequestInt = nil
+            if len(listWTimeout) > 0 {
+                r = listWTimeout[0].req
+                listWTimeout = listWTimeout[1:]
+            } else if len(listNoTimeout) > 0 {
+                r = listNoTimeout[0].req
+                listNoTimeout = listNoTimeout[1:]
+            }
+            if r != nil {
+                r.ChResponse <- &LoMResponse { 0, "", serverRequests[0] }
+                serverRequests = serverRequests[1:]
+            } else {
+                break
+            }
+        }
+
+        tnow := time.Now().Unix()
+        tout = A_DAY_IN_SECS
+        for i := 0; i < len(listWTimeout); i++ {
+            if tnow >= listWTimeout[i].due {
+                /* Fail the request */
+                r := listWTimeout[i].req
+                r.ChResponse <- &LoMResponse { int(LoMReqTimeout),
+                                GetLoMResponseStr(LoMReqTimeout), nil }
+            } else {
+                /* Get tout and drop the failed ones */
+                tout = 0
+                if tnow < listWTimeout[i].due {
+                    tout = listWTimeout[i].due - tnow
+                }
+                listWTimeout = listWTimeout[i:]
+                break
+            }
+        }
+
+    }
+}
+
+
+/*
+ * A go func watches, yet it may get blocked by actions like publish
+ * To ensure, notify APIs are not blocked, have a buffer.
+ * Publish API is generally pretty quick, a value of 10 is more than enough.
+ */
+const HEARTBEAT_CH_SIZE = 10
+
+/* All registrations & heartbeats from clients. */
+type ClientRegistrations_t struct {
+    activeActions   map[string]*ActiveActionInfo_t
+    activeClients   map[string]*ActiveClientInfo_t
+    heartbeatCh     chan string     /* Action name */
+}
+
+/* Initialized object; But not exported */
+var clientRegistrations *ClientRegistrations_t = nil
+
+
+func GetRegistrations() *ClientRegistrations_t {
+    if clientRegistrations == nil {
+        InitRegistrations()
```

Comment:
Is engine single threaded ? Will this method always be called from a single goroutine so that two different clientRegistration instances are not returned to two different callers ?

Reply:
Yes. All calls are from single loop in  engine::runLoop